libsawtooth/src/receipt/store/diesel/operations/add_txn_receipts.rs

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright 2021 Cargill Incorporated
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ------------------------------------------------------------------------------
+ */
+
+//! Provides the add transaction receipts operation for `DieselReceiptStore`
+
+use std::convert::TryFrom;
+
+use diesel::{insert_into, prelude::*};
+use transact::protocol::receipt::{StateChange, TransactionReceipt, TransactionResult};
+
+use crate::error::InternalError;
+
+use super::ReceiptStoreOperations;
+
+use crate::receipt::store::{
+    diesel::{
+        models::{
+            InvalidTransactionResultModel, NewValidTransactionResultDataModel,
+            NewValidTransactionResultEventModel, NewValidTransactionResultStateChangeModel,
+            StateChangeTypeModel, TransactionReceiptModel,
+            ValidTransactionResultEventAttributeModel, ValidTransactionResultEventModel,
+        },
+        schema::{
+            invalid_transaction_result, transaction_receipt, valid_transaction_result_data,
+            valid_transaction_result_event, valid_transaction_result_event_attribute,
+            valid_transaction_result_state_change,
+        },
+    },
+    ReceiptStoreError,
+};
+
+pub(in crate::receipt::store::diesel) trait ReceiptStoreAddTxnReceiptsOperation {
+    fn add_txn_receipts(&self, receipts: Vec<TransactionReceipt>) -> Result<(), ReceiptStoreError>;
+}
+
+impl<'a> ReceiptStoreAddTxnReceiptsOperation
+    for ReceiptStoreOperations<'a, diesel::sqlite::SqliteConnection>
+{
+    fn add_txn_receipts(&self, receipts: Vec<TransactionReceipt>) -> Result<(), ReceiptStoreError> {
+        self.conn.transaction::<(), _, _>(|| {
+            for (index, receipt) in receipts.iter().enumerate() {
+                if transaction_receipt::table
+                    .filter(transaction_receipt::transaction_id.eq(receipt.transaction_id.to_string()))
+                    .first::<TransactionReceiptModel>(self.conn)
+                    .optional()?
+                    .is_some()
+                {
+                    return Err(ReceiptStoreError::InternalError(InternalError::with_message(format!(
```

Comment:
This should probably return a `ConstraintViolationError`

Reply:
Updated to be a ConstraintViolationError